keps/sig-node/20190322-image-gc-whitelist.md

```diff
@@ -0,0 +1,305 @@
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives [optional]](#alternatives-optional)
+  - [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Enhancing Kubelet image garbage collector with a whitelist of container images. Whitelisted images will be ignored during garbage collection.
+
+## Motivation
+
+There are some use cases when not all the container images are available from a remote repository, therefore container images are stored locally in the image cache of the Nodes. These container images can be preloaded to the Nodes instead of pulled on demand.
```

Comment:
Could you share more about the use cases in which all the container images are not available from a remote repository? I certainly understand not wanting to host certain images on a public repository, but I'm not sure I see a time when it would make sense to avoid image repositories entirely, especially when we would need to add extra complexity to the Kubelet to compensate.

Reply:
@mattjmcnaughton let me explain our use-case. We deliver our VNF products to on-prem, isolated, private customer environments, where the K8S cluster is embedded in the product. Naturally, the business logic application container images are hosted in an embedded container registry to have proper management for it (scaling, upgrade, etc.).
The container infra however, needs different approach to manage in our scenario. The nodes are 100% standalone from deployment PoV, there is no central source of content (like container images, RPMs, etc.). Everything is packaged inside a single deliverable (i.e. qcow disk image) which is able to spawn up every node according to it's role on IaaS clouds or BM hosts. This is convenient from integration and from user PoV. It is also very fast, because for example we don't have network/disk shortage during deployment due to 70 nodes are pulling down the container infra images from the same private registry. The upgrade is also performed on node level (from new qcow) for similar reasons. This way, there is no need to have special, grand preparation before, or garbage collection after all the nodes are upgraded.
When the nodes are such self-containing, there is need to exclude the containerized parts of the infrastructure from accidental garbage collection. So here we are :smile:

Reply:
Interesting interesting :) Thanks for sharing more! Very interested to hear from the community on this, re if other people have a similar use case and how they are currently solving it (if at all).

Reply:
this would be awesome in situations where you can't afford to lose time between pulling an image again and getting it up and running if your pods die.
I've been in situations where I've had to disable the cleanup , so that some of the images don't get wiped. 🔥

Reply:
@dashpole i just max the vars out which is silly, and not actually disabling it

Reply:
@dashpole for example with Core OS Container Linux, you have toolchain, which loads a container and emergency cases you need this to load, because CoreOS just have not the tools required. I would liked to pull my toolchain container and have it whitelisted.

Reply:
[kind](https://github.com/kubernetes-sigs/kind) additionally works this way, when you build & test kubernetes locally the core kubernetes images are never pushed to a registry (you wouldn't have access) they're just side-loaded into the container runtime on each of the nodes.

currently we turn off image GC and disk eviction because of this, these are short lived non-production clusters, but it is at least one use case for having images that should never be evicted :-)